### PR TITLE
pkg/bpf: add sync.Mutex to Map type

### PR DESCRIPTION
**Pull request by ianvernon**
Utilize lock for all map operations to serialise access since the system calls do not do this on their own.

Signed-off by: Ian Vernon <ian@covalent.io>

### CODE REVIEW

File: pkg/bpf/map.go [hunk 1]
```diff
@@ -309,11 +320,13 @@ func (m *Map) Dump(parser DumpParser, cb DumpCallback) error {
 
 		copy(key, nextKey)
 	}
-
 	return nil
 }
 
 func (m *Map) Lookup(key MapKey) (MapValue, error) {
+	m.lock.Lock()
```
Comment (tgraf) [inline pkg/bpf/map.go]: Can this be an RLock()?

**Review by tgraf — COMMENTED** [review]
Do we have existing code with locking that can be removed now that we have this in the generic layer?

File: pkg/bpf/map.go (hunk 1 above)
Comment (ianvernon) [reply pkg/bpf/map.go]: @aanm said the same thing :) sure - I will do the same for Dump as well.

**Comment by ianvernon** [comment]
I will check to see if there is locking code that can be removed now that we have it integrated into the map itself.

**Comment by ianvernon** [comment]
I went through the existing code that uses the map functions (e.g., loadbalancer.go), and the locks that cover a lot of the calls (Dump, Add, etc.) are needed for other constructs besides the BPF map itself, so I did not find any that could be removed.

**Comment by aanm** [comment]
What about `pkg/maps/lxcmap/lxcmap.go:L37`? Do we still need that lock?

**Comment by ianvernon** [comment]
We won't need that lock once we convert lxcmap to use the pkg/bpf/map.go map interface. Currently the calls to update the maps are to UpdateElement, DeleteElement, etc. which do not have synchronization, while the map interface functions (Update, Delete) lock the map and call the UpdateElement and DeleteElement functions. So until #357 is resolved we cannot remove this mutex.

**Review by tgraf — APPROVED** [review]

**Pull request merged**